src/txmempool.cpp

```diff
@@ -57,48 +57,70 @@ size_t CTxMemPoolEntry::GetTxSize() const
 // Update the given tx for any in-mempool descendants.
 // Assumes that setMemPoolChildren is correct for the given tx and all
 // descendants.
-void CTxMemPool::UpdateForDescendants(txiter updateIt, cacheMap &cachedDescendants, const std::set<uint256> &setExclude)
+void CTxMemPool::UpdateForDescendants(txiter update_it, cacheMap& cache, const std::set<uint256>& exclude)
 {
-    setEntries stageEntries, setAllDescendants;
-    stageEntries = GetMemPoolChildren(updateIt);
-
-    while (!stageEntries.empty()) {
-        const txiter cit = *stageEntries.begin();
-        setAllDescendants.insert(cit);
-        stageEntries.erase(cit);
-        const setEntries &setChildren = GetMemPoolChildren(cit);
-        for (txiter childEntry : setChildren) {
-            cacheMap::iterator cacheIt = cachedDescendants.find(childEntry);
-            if (cacheIt != cachedDescendants.end()) {
-                // We've already calculated this one, just add the entries for this set
-                // but don't traverse again.
-                for (txiter cacheEntry : cacheIt->second) {
-                    setAllDescendants.insert(cacheEntry);
+    const auto epoch = GetFreshEpoch();
+    const CTxMemPool::setEntries& direct_children = GetMemPoolChildren(update_it);
+    // set up the update_cache to contain all of our transaction's children (note --
+    // already de-duplicated in case multiple outputs of ours are spent in one
+    // transaction)
+    vecEntries update_cache;
+    update_cache.reserve(direct_children.size());
+    // mark every direct_child as visited so that we don't accidentally re-add them
+    // to the cache in the grandchild is child case
```

Comment:
nit: This comment could be improved, for code readers not as familiar with the issues here.  Perhaps something like:
```
mark every direct child as visited, so that if we encounter one of these transactions again 
(as the descendant of some other direct child), we don't re-add them to the cache.
```

Reply:
Maybe adding an example helps?

```
mark every direct child as visited, so that if we encounter one of these transactions again 
(as the descendant of some other direct child), we don't re-add them to the cache. e.g.,

Tx A: L1 -> {O1, O2}
Tx B: {O1} -> {P1}
Tx C: {P1, O2} -> {Q1}

In this case transaction C is both a direct child of A and a grandchild via B.
```

I can improve the documentation with this in the follow up PR where we clean this up more (because maybe we'd just end up removing the caching anyways).